File: from-0101-to-0200/0141-linked-list-cycle/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return False
        
        p_slow = head
        p_fast = head
        
        while p_fast.next:
            p_fast = p_fast.next
            if p_fast == p_slow:
                return True
            if p_fast.next:
                p_fast = p_fast.next
            else:
                return False            
            if p_fast == p_slow:
                return True
            p_slow = p_slow.next
            
        return False

File: from-0101-to-0200/0141-linked-list-cycle/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func hasCycle(head *ListNode) bool {
    if head == nil || head.Next == nil {
        return false
    }

    fast, slow := head, head
    for fast.Next != nil {
        fast = fast.Next
        if fast == slow {
            return true
        }
        fast = fast.Next
        if fast == nil {
            break
        }
        if fast == slow {
            return true
        }
        slow = slow.Next
    }
    return false
}